File: elixir/bob/bob.exs
defmodule Bob do
  @spec hey(String.t) :: String.t
  def hey(input) do
    is_question = String.ends_with?(String.trim_trailing(input), "?")
    is_empty = String.trim(input) == ""
    # Only consider the input to be uppercase if it's uppercase and if
    # there are characters that are effected by casing.
    is_uppercase = String.upcase(input) == input && String.downcase(input) != input

    cond do
      is_empty ->
        # Not saying anything.
        "Fine. Be that way!"
      is_question && is_uppercase ->
        # Yelling a question.
        "Calm down, I know what I'm doing!"
      is_uppercase ->
        # Yelling something other than a question.
        "Whoa, chill out!"
      is_question ->
        # Asking a question.
        "Sure."
      true ->
        # Everything else.
        "Whatever."
    end
  end
end

File: erlang/bob/src/bob.erl
-module(bob).

-export([response/1, test_version/0]).

%% API

-spec response(unicode:chardata()) -> unicode:chardata().
response(String) ->
    IsQuestion = is_question(String),
    IsEmpty = is_empty(String),
    IsUppercase = is_uppercase(String),

    if
        IsEmpty ->
            % Not saying anything
            "Fine. Be that way!";
        IsQuestion andalso IsUppercase ->
            % Yelling a question.
            "Calm down, I know what I'm doing!";
        IsUppercase ->
            % Yelling something other than a question.
            "Whoa, chill out!";
        IsQuestion ->
            % Asking a question.
            "Sure.";
        true ->
            % Everything else.
            "Whatever."
    end.


-spec test_version() -> integer().
test_version() ->
    3.

%% Internal

is_question(String) ->
    Trimmed = string:trim(String, trailing),
    string:find(Trimmed, "?", trailing) == "?".

is_empty(String) ->
    string:trim(String) == "".

is_uppercase(String) ->
    string:uppercase(String) == String andalso
        string:lowercase(String) /= String.
